Allowing Remotesync RPC Return Values

AsherGlick opened the pull request:
Implementation of proposal https://github.com/godotengine/godot-proposals/issues/879

Does not change the return type of the gdscript functions `rpc` or `rpc_id` but does cause some previously empty return values to become non-empty.

AsherGlick committed:
Add a return value for rpc() and rpc_id() when calling a *sync function

When using a `remotesync`, `mastersync`, or `puppetsync` function via `rpc` or
`rpc_id`, that function is also run on the local machine, however there is no
way to retrieve the return value from that local function run.

This patch allows the return value to be used without converting the rpc into
a `remote`, `master`, or `puppet` function and calling it via rpc and manually

Implementation of godotengine/godot-proposals#879

AsherGlick commented:
If this PR gets approved I have a separate branch for 3.2 and would like some guidance on if I should also submit the 3.2 backport of the feature or not.

akien-mga commented:
Please amend the commit to use a meaningful description: https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-messages-with-readability-in-mind

Faless commented:
I think this will be confusing for users, since the result will always be from the local call.
I'm not convinced that we actually need this change.

AsherGlick commented:
@akien-mga Thank you for the link. I have updated the commit message to hopefully be more descriptive.

@Faless I am concerned about it being a bit confusing as well. I think the primary confusion I worry about would be a new developer asking "why is my `remote` function not returning a value when I call it via `rpc()`" followed by someone trying to help suggesting to change the function def from `remote` to `remotesync` and then the developer assuming `*sync` means "wait for a return" and not "run on my local machine as well". However once the new developer has a simple understanding of what `sync` means I believe this confusion would be eliminated.

I have updated my proposal with https://github.com/godotengine/godot-proposals/issues/879#issuecomment-633664707 showing a practical example of why this would be useful in my own work.

AsherGlick commented:
I am going to close this pull request as I agree now that it is too confusing given the non obvious scenarios of a "master" function being called on a server, or a "puppet" function being called on a client and how that would interact with the return value, as discussed in the proposal.

Thank you everyone involved for your time with this pull request.
